domain/removal/state/unit.go

```diff
@@ -65,62 +68,101 @@ AND    life_id = 0`, unitUUID)
-With machines AS (
-	SELECT    m.uuid AS machine_uuid,
-	          u.uuid AS unit_uuid,
-			  COUNT(u.uuid) AS unit_count
-	FROM      machine AS m
-	JOIN      net_node AS nn ON nn.uuid = m.net_node_uuid
-	LEFT JOIN unit AS u ON u.net_node_uuid = nn.uuid
-	GROUP BY  m.uuid
+WITH units_alive AS (
+    SELECT uuid, net_node_uuid
+    FROM   unit
+    WHERE  life_id = 0
+), units_not_alive AS (
+    SELECT uuid, net_node_uuid
+    FROM   unit
+    WHERE  life_id != 0
+), machines AS (
+    SELECT    m.uuid AS machine_uuid,
+              m.net_node_uuid,
+              COUNT(ua.uuid) AS unit_alive_count,
+              COUNT(una.uuid) AS unit_not_alive_count
+    FROM      machine AS m
+    JOIN      net_node AS nn ON nn.uuid = m.net_node_uuid
+    LEFT JOIN units_alive AS ua ON ua.net_node_uuid = nn.uuid
+    LEFT JOIN units_not_alive AS una ON una.net_node_uuid = nn.uuid
+    GROUP BY  m.uuid
 )
-SELECT unit_count AS &entityAssociationCount.count,
-	   machine_uuid AS &entityAssociationCount.uuid
+SELECT unit_alive_count AS &entityAssociationAliveCount.alive_count,
+       unit_not_alive_count AS &entityAssociationAliveCount.not_alive_count,
+       machine_uuid AS &entityAssociationAliveCount.uuid
 FROM   machines
-WHERE  unit_uuid = $entityUUID.uuid;
-	`, unitUUID, entityAssociationCount{})
+LEFT JOIN unit AS u ON u.net_node_uuid = machines.net_node_uuid
+WHERE  u.uuid = $entityUUID.uuid;
+    `, unitUUID, entityAssociationAliveCount{})
 	if err != nil {
 		return "", errors.Errorf("preparing unit count query: %w", err)
 	}
 
 	updateMachineStmt, err := st.Prepare(`
 UPDATE machine
 SET    life_id = 1
-WHERE  uuid = $entityAssociationCount.uuid
-AND    life_id = 0`, entityAssociationCount{})
+WHERE  uuid = $entityUUID.uuid
+AND    life_id = 0`, entityUUID{})
 	if err != nil {
 		return "", errors.Errorf("preparing machine life update: %w", err)
 	}
 
-	var mUUID string
-	if err := errors.Capture(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		if err := tx.Query(ctx, updateUnitStmt, unitUUID).Run(); err != nil {
-			return errors.Errorf("advancing unit life: %w", err)
-		}
-
-		var unitCount entityAssociationCount
-		if err := tx.Query(ctx, lastUnitStmt, unitUUID).Get(&unitCount); errors.Is(err, sqlair.ErrNoRows) {
-			return nil
-		} else if err != nil {
-			return errors.Errorf("getting unit count: %w", err)
-		} else if unitCount.Count != 1 {
-			// The unit is not the last one on the machine.
-			return nil
-		}
-
-		if err := tx.Query(ctx, updateMachineStmt, unitCount).Run(); err != nil {
-			return errors.Errorf("advancing machine life: %w", err)
-		}
+	var unitCount entityAssociationAliveCount
+	if err := tx.Query(ctx, lastUnitStmt, unitUUID).Get(&unitCount); errors.Is(err, sqlair.ErrNoRows) {
+		return "", nil
+	} else if err != nil {
+		return "", errors.Errorf("getting unit count: %w", err)
+	} else if unitCount.AliveCount > 0 {
+		// Nothing to do.
+		return "", nil
+	} else if unitCount.NotAliveCount == 0 {
```

Comment:
This is the inverse of the alive count. You don't need it here or in the query. Check for alive units, if there are any, do nothing, otherwise the machine goes too.

Reply:
The intent is to perform any additional busy work attempting to set the machine as dying if the life is alive.

Prerequisites:

 - The unit we're deleting will already be set to dying.
 - The machine will still be alive.

If there are any units on the machine; then we must ensure that non of them are in the dying phase. Yet, if there are units that exist on the machine that are already set to dying we should also set it to dying.

Scenario:

 - Add two units to a machine.
 - Delete first unit
 - Delete second unit
 - Should transition machine to dying.

The second guard `NotAliveCount` states that if there is nothing on this machine that is either alive, dying or dead, don't touch it. That was the original hypothesis. The question becomes, is it a superfluous check?

Reply:
If we remove the unit_not_alive check, the tests fail because of the scenario.

Reply:
I see. The second part works under the assumption that the killer of the last unit will have set it to dying. I'm trying to think how this goes if the machine was previously prevented from dying due to having containers that it no longer does. I think we can think about that later.